Constraint and regularization in Embedding layer

vzhong opened the pull request:
I've noticed that as of its current implementation, the Embedding layer doesn't support constraints and regularization. One type of constraint that we have found to be empirically useful (eg. better generalization, sometimes faster convergence) is constraining the embeddings to have unit norm. This technique is sometimes mentioned in embedding approaches [[1](http://www.aclweb.org/anthology/N13-1090), [2](http://www.cs.toronto.edu/~zemel/documents/nips2014_kiros.pdf)].

I've made an attempt at implementing this here. Please let me know what you think :)

Moreover, if you'd like me to make a finer grain unit test on `unitnorm` (eg. via Theano `function`), I can do that as well. Overall I think it may be nice to migrate to fine grain unit tests across the board with python `unittest` and only keep around a few integration tests. That way developers have an easier time checking against regression (eg. by running `python -m unittest discover`.

fchollet commented:
Looking good! Would you mind updating the Embedding layer documentation as well?

Agreed that the unit testing process needs to be normalized...

fchollet commented:
Also, one question: have you had good results with the unitnorm constraint applied to the Embedding layer on any task? Specifically, better results than with no constraints?

vzhong commented:
Updated docs! Also I removed a unicode dash in one of the comments (probably came from a copy and paste somewhere), hope you don't mind :)

Regarding unit norm constraints: we use them in the task of slot filling. That is, given a entity and a relation, we'd like to predict the other entity (eg. `Barack_Obama per:born_in <fill>`). In particular, having unit norm constraints on the entity and relation representations helps the model generalize. I can't say whether it helps with convergence as well at this point. This is actually ongoing work. At some point, I may be able to push a small example of doing things like this with a small dataset.
